juju/juju
# Fix azure instance queries to account for old completed deployments

Using a custom Azure resource group, you can bootstrap and then destroy the controller and re-use the group for the next bootstrap. However, Juju had a buggy method of counting machines in a resource group - it would query the completed Deployment records. Because these records are retained on the group, the next bootstrap would think there are too many machines and exit.

This PR changes how machines are queried - it still has to look for deployments because there's a window where the deployment is created but the machine has not shown up yet. So Juju will now query any _incomplete_  deployments, and augment that list with any machines as well. Any machines found take precedence over deployments. The method to gather instances uses the same logic as for EC2 - a short retry loop to allow the requested instances time to get created.

Also fix some lint issues and the resources API version needed to be bumped.

## QA steps

bootstrap to a previously created resource group, add a machine to get 2 azure deployments recorded, destroy the controller, manually delete the resource group contents, and bootstrap again
```plain
juju bootstrap --config resource-group-name=test --no-default-model azure/eastus aztest
juju add-machine
juju destroy controller aztest
```
use the azure portal to see there's 2 Deployments recorded; manually delete resource group content
bootstrap again
`juju bootstrap --config resource-group-name=test --no-default-model azure/eastus`

Previously this would fail as per the bug.

## Bug reference

https://bugs.launchpad.net/juju/+bug/1935979
